webknossos/webknossos/dataset/_array.py

```diff
@@ -296,6 +301,466 @@ def __setstate__(self, d: Dict[str, Any]) -> None:
+
+    def __init__(self, credentials_file_path: Path) -> None:
+        self.entries = {}
+        self.credentials_file_path = credentials_file_path
+
+    def _dump_credentials(self) -> None:
+        self.credentials_file_path.write_text(
+            "\n".join(
+                [
+                    f"[profile-{key_hash}]\naws_access_key_id = {access_key_id}\naws_secret_access_key = {secret_access_key}\n"
+                    for key_hash, (
+                        access_key_id,
+                        secret_access_key,
+                    ) in self.entries.items()
+                ]
+            )
+        )
+
+    def add(self, access_key_id: str, secret_access_key: str) -> Dict[str, str]:
+        key_tuple = (access_key_id, secret_access_key)
+        key_hash = hash(key_tuple)
+        self.entries[key_hash] = key_tuple
+        self._dump_credentials()
+        return {
+            "profile": f"profile-{key_hash}",
+            "filename": str(self.credentials_file_path),
+        }
+
+
+@lru_cache
+def _aws_credential_file() -> Path:
+    temp_dir_path = Path(mkdtemp())
+    credentials_file_path = Path(temp_dir_path / "aws_credentials")
+    credentials_file_path.touch()
+    return credentials_file_path
+
+
+_aws_credential_manager = AWSCredentialManager(_aws_credential_file())
+
+
+class TensorStoreArray(BaseArray):
+    _cached_array: Optional[tensorstore.TensorStore]
+
+    def __init__(
+        self, path: Path, _cached_array: Optional[tensorstore.TensorStore] = None
+    ):
+        super().__init__(path)
+        self._cached_array = _cached_array
+
+    @staticmethod
+    def _make_kvstore(path: Path) -> Union[str, Dict[str, Union[str, List[str]]]]:
+        if is_fs_path(path):
+            return {"driver": "file", "path": str(path)}
+        elif isinstance(path, UPath) and path.protocol in ("http", "https"):
+            return {
+                "driver": "http",
+                "base_url": str(path),
+                "headers": [
+                    f"{key}: {value}"
+                    for key, value in path.storage_options.get("headers", {}).items()
+                ],
+            }
+        elif isinstance(path, UPath) and path.protocol in ("s3"):
+            parsed_url = urlparse(str(path))
+            kvstore_spec: dict[str, Any] = {
+                "driver": "s3",
+                "path": parsed_url.path,
+                "bucket": parsed_url.netloc,
+            }
+            if endpoint_url := path.storage_options.get("client_kwargs", {}).get(
+                "endpoint_url", None
+            ):
+                kvstore_spec["endpoint"] = endpoint_url
+            if "key" in path.storage_options and "secret" in path.storage_options:
+                kvstore_spec["aws_credentials"] = _aws_credential_manager.add(
+                    path.storage_options["key"], path.storage_options["secret"]
+                )
+            return kvstore_spec
+        else:
+            return str(path)
```

Comment:
Is a string enough for a KVStoreSpec? I would have assumed that something like `{"driver": "file", "path": str(path)}` is required here.